draft-ietf-core-problem-details.md

```diff
@@ -13,224 +17,428 @@ stand_alone: yes
+machine-readable details of errors in a REST response to avoid the
+need to define new error response formats for REST APIs.
+The format
+is inspired by, but intended to be more concise than, the Problem
+Details for HTTP APIs defined in RFC 7807.
 
 --- middle
 
 # Introduction
 
-CoAP {{!RFC7252}} response codes are sometimes not sufficient to convey enough information about an error to be helpful.  This specification defines a simple and extensible CoRAL {{!I-D.ietf-core-coral}} vocabulary to suit this purpose.  It is designed to be reused by CoAP APIs, which can identify distinct "problem types" specific to their needs.  Thus, API clients can be informed of both the high-level error class (using the response code) and the finer-grained details of the problem (using this vocabulary), as shown in {{fig-problem-details}}.
-
-~~~
-{::include problem-details.ascii-art}
+REST response status information such as CoAP response
+codes ({{Section 5.9 of -coap}}) is sometimes not sufficient to convey enough information about
+an error to be helpful.  This specification defines a simple and extensible
+framework to define CBOR {{-cbor}} data items to suit this purpose.
+It is designed to be reused by REST APIs, which can identify distinct
+"problem types" specific to their needs.
+Thus, API clients can be informed of both the high-level error class
+(using the response code) and the finer-grained details of the problem
+(using this vocabulary), as shown in {{fig-problem-details}}.
+
+~~~ aasvg
++--------+           +--------+
+|  CoAP  |           |  CoAP  |
+| Client |           | Server |
++----+---+           +----+---+
+     |                    |
+     | Request            |
+     |------------------> |
+     |                    |
+     | <----------------- |
+     | Error Response     |
+     | with a CBOR Data   |
+     | Item giving        |
+     | Problem Details    |
+     |                    |
 ~~~
-{: #fig-problem-details artwork-align="center" title="Problem Details"}
+{: #fig-problem-details artwork-align="center"
+   title="Problem Details: Example with CoAP"}
 
-The vocabulary presented is largely inspired by the Problem Details for HTTP APIs defined in {{?RFC7807}}.
+The framework presented is largely inspired by the Problem Details for HTTP APIs defined in {{RFC7807}}.
 
 ## Requirements Language
 
-{::boilerplate bcp14}
+{::boilerplate bcp14-tagged}
 
 # Basic Problem Details
 
-A CoRE Problem Details is a CoRAL document with the following elements:
+A Concise Problem Details data item is a CBOR data item with the following
+structure (notated in CDDL {{!RFC8610}}):
+
+~~~ CDDL
+problem-details = non-empty<{
+  ? &(title: -1) => text
+  ? &(detail: -2) => text
+  ? &(instance: -3) => ~uri
+  ? &(response-code: -4) => uint .size 1
+  standard-problem-detail-entries
+  custom-problem-detail-entries
+}>
+
+standard-problem-detail-entries = (
+  * nint => any
+)
+
+custom-problem-detail-entries = (
+  * (uint/~uri) => { + any => any }
```

Comment:
If we allow URIs in here, I'm afraid we're back to where we can't easily paint a way toward CoRAL convertability (see also my comments during the last interim).

Reply:
If that's just about replacing private-use uint numbers, that's fine with me (because there is no expectation of convertability with private-use numbers either); maybe I missed it or maybe it can be phrased more explicitly.